src/hotspot/share/runtime/thread.cpp

```diff
@@ -2543,17 +2544,33 @@ void JavaThread::verify() {
 // seen prior to having its threadObj set (e.g., JNI attaching threads and
 // if vm exit occurs during initialization). These cases can all be accounted
 // for such that this method never returns NULL.
-const char* JavaThread::get_thread_name(const char* default_name) const {
-  Thread* thread = Thread::current();
-  ThreadsListHandle tlh(thread);
-
-  if (thread->is_JavaThread_protected(this)) {
+const char* JavaThread::get_thread_name() const {
+  if (Thread::is_JavaThread_protected(this)) {
     // The target JavaThread is protected so get_thread_name_string() is safe:
     return get_thread_name_string();
   }
 
+//#ifdef ASSERT
+  Thread *current_thread = Thread::current();
+//#endif
+  // Use this with -XX:+UseNewCode to diagnose locations that are missing
+  // a ThreadsListHandle:
+  // guarantee(!UseNewCode, "current_thread=" INTPTR_FORMAT
+  //           " is not protecting this=" INTPTR_FORMAT,
+  //           p2i(current_thread), p2i(this));
+  guarantee(false, "current_thread=" INTPTR_FORMAT
+            " is not protecting this=" INTPTR_FORMAT,
+            p2i(current_thread), p2i(this));
+
+  // Note: Since this JavaThread isn't protected by a TLH, the call to
+  // this->is_handshake_safe_for() may crash, but we have debug bits so...
+  assert(SafepointSynchronize::is_at_safepoint() ||
+         this->is_handshake_safe_for(current_thread), "JavaThread="
```

Comment:
I agree this is a pre-existing bug as accessing `this` may not be safe if the target is not protected. But if we crash attempting the access that implies the assert failed so ...

Reply:
... that will tell us that we have a spot in the code where a TLH is
missing once we've analyzed the crash. So are you okay with this
assert() having a bug since the original assert has the same bug?

Reply:
It has occurred to me that I should have moved all of the
protection checking logic and assertion checks into the
new Thread::is_JavaThread_protected() function. That will
allow JavaThread::get_thread_name() to become very simple.